File: alpha/js/__tests__/testNamespace.js
var Namespace = require('../es6/namespace.js');
var Experiment = require('../es6/experiment.js');

var global_log = [];
class Experiment1 extends Experiment {
  configure_logger() {
    return;
  }

  log(data) {
    global_log.push(data);
  }

  previously_logged() {
    return true;
  }

  setup() {
    this.name = 'test_name';
  }

  assign(params, args) {
    params.set('test', 1)
  }
}

class Experiment2 extends Experiment {
  configure_logger() {
    return;
  }

  setup() {
    this.name = 'test_name';
  }

  log(data) {
    global_log.push(data);
  }

  previously_logged() {
    return true;
  }

  assign(params, args) {
    params.set('test', 2)
  }
}


describe("Test namespace module", function() {
  var validate_log;
  beforeEach(function() {
    validate_log = function(exp) {
      expect(global_log[0].salt).toEqual(`test-${exp}`)
    }

  });

  afterEach(function() {
    global_log = [];
  });
  it('Adds segment correctly', function() {
    class TestNamespace extends Namespace.SimpleNamespace {
      setup() {
        this.name = "test";
        this.num_segments = 100;
        this.set_primary_unit('userid');
      }

      setup_experiments() {
        this.add_experiment('Experiment1', Experiment1, 100);
      }
    }
    var namespace = new TestNamespace({'userid': 'blah'});
    expect(namespace.get('test')).toEqual(1);
    validate_log("Experiment1");
  });

  it('Adds two segments correctly', function() {
    class TestNamespace extends Namespace.SimpleNamespace {
      setup() {
        this.name = "test";
        this.num_segments = 100;
        this.set_primary_unit('userid');
      }

      setup_experiments() {
        this.add_experiment('Experiment1', Experiment1, 50);
        this.add_experiment('Experiment2', Experiment2, 50);
      }
    }
    var namespace = new TestNamespace({'userid': 'blah'});
    expect(namespace.get('test')).toEqual(1);
    validate_log("Experiment1");
    global_log = [];
    var namespace2 = new TestNamespace({'userid': 'abb'});
    expect(namespace2.get('test')).toEqual(2);
    validate_log("Experiment2");
  });

  it('Can remove segment correctly', function() {
    class TestNamespace extends Namespace.SimpleNamespace {
      setup() {
        this.name = "test";
        this.num_segments = 10;
        this.set_primary_unit('userid');
      }

      setup_experiments() {
        this.add_experiment('Experiment1', Experiment1, 10);
        this.remove_experiment('Experiment1');
        this.add_experiment('Experiment2', Experiment2, 10);
      }
    }
    var str = "bla";
    for(var i = 0; i < 100; i++) {
      str += "h";
      var namespace = new TestNamespace({'userid': str});
      expect(namespace.get('test')).toEqual(2);
      validate_log("Experiment2");
    }

  });
});

File: python/planout/test/test_namespace.py
import unittest

from planout.namespace import SimpleNamespace
from planout.experiment import DefaultExperiment


class VanillaExperiment(DefaultExperiment):
    def setup(self):
        self.name = 'test_name'

    def assign(self, params, i):
        params.foo = 'bar'


class DefaultExperiment(DefaultExperiment):
    def get_default_params(self):
        return {'foo': 'default'}


class NamespaceTest(unittest.TestCase):
    def test_namespace_remove_experiment(self):
        class TestVanillaNamespace(SimpleNamespace):
            def setup(self):
                self.name = 'test_namespace'
                self.primary_unit = 'i'
                self.num_segments = 100
                self.default_experiment_class = DefaultExperiment

            def setup_experiments(self):
                self.add_experiment('test_name', VanillaExperiment, 100)
                self.remove_experiment('test_name')

        assert TestVanillaNamespace(i=1).get('foo') == 'default'

    def test_namespace_add_experiment(self):
        class TestVanillaNamespace(SimpleNamespace):
            def setup(self):
                self.name = 'test_namespace'
                self.primary_unit = 'i'
                self.num_segments = 100
                self.default_experiment_class = DefaultExperiment

            def setup_experiments(self):
                self.add_experiment('test_name', VanillaExperiment, 100)

        assert TestVanillaNamespace(i=1).get('foo') == 'bar'


if __name__ == '__main__':
    unittest.main()
